In [4]:
# Các gói cần dùng
import numpy as np
import statsmodels
import scipy.stats
import pandas as pd

### I. Kiểm định một trung bình

In [1]:
# xem chi tiết các tham số
from scipy.stats import ttest_1samp
?ttest_1samp

In [2]:
ttest_1samp # xem tên các tham số

<function scipy.stats._stats_py.ttest_1samp(a, popmean, axis=0, nan_policy='propagate', alternative='two-sided', *, keepdims=False)>

#### Hướng dẫn cách dùng tham số:

scipy.stats.ttest_1samp(a, popmean, nan_policy='propagate', 
                        alternative='two-sided')
                        
Trong đó, a: array-like chứa dữ liệu mẫu

popmean: Giá trị mu_0 trong H0

nan_policy : {'propagate', 'omit', 'raise'} cách thức xử lý số liệu trống

    + propagate: không tính toán, trả ra NaN
    
    + omit: NaN được loại bỏ trước khi tính toán
    
    + raise: Một cách thức tính khác 
    
Nếu khi có số liệu trống trong bài, ta sẽ sử dụng 'omit'

alternative : {'two-sided', 'less', 'greater'}: mô tả giả thuyết đối

    + two-sided: hai bên
    
    + less: bên trái
    
    + greater: bên phải

Bai 16: Chiều cao trung bình của nam giới của một nước được cho là 1.78
m. Để kiểm tra chiều cao của nam giới tại một thành phố có ở mức trung bình như
trên toàn quốc hay không, một mẫu ngẫu nhiên gồm 20 người là nam giới được chọn
ra. Chiều cao của họ được ghi lại trong file ChieuCao.csv. Cho biết chiều cao nam giới tại thành phố này tuân theo phân phối chuẩn, hãy kết
luận ở mức ý nghĩa α = 1%.

In [5]:
ChieuCao=pd.read_csv('ChieuCao.csv')

In [6]:
print(ChieuCao.head())

   TT  ChieuCao
0   1      1.83
1   2      1.73
2   3      1.76
3   4      1.85
4   5      1.81


Gọi m là chiều cao trung bình của nam giới ở vùng này.

H0: m = 1.78         

H1: m # 178

p-giá trị = 0.006 < 0.01 => Bác bỏ H0

=> Ở mức ý nghĩa 1% ta kết luận: Chiều cao trung bình của nam giới ở vùng này không giống như trên toàn quốc.

In [7]:
# Lệnh thực hiện tìm p-giá trị
ttest_1samp(ChieuCao.ChieuCao,popmean=1.78,alternative='two-sided')

TtestResult(statistic=3.085293839515478, pvalue=0.006091466065562459, df=19)

Trong kết quả này, 
+ statistic cho biết giá trị kiểm định t, 
+ pvalue: cho biết p - giá trị
+ df: cho biết bậc tự do của thống kê t

Bài 19: Một công ty sản xuất những tấm nhựa dùng trong công nghiêp vừa
đưa ra một sản phẩm mới và cho rằng sản phẩm này có thể chịu được sức ép trung
bình ít nhất là 30 (psi), ở đây khả năng chịu lực được tính bằng số pound cần thiết
làm vỡ tấm nhựa trên một diện tích 1 inch vuông. Dữ liệu sau đây cho biết khả
năng chịu lực của một mẫu gồm 16 tấm nhựa (file dữ liệu Nhua.txt). Giả thiết rằng khả năng chịu lực của các tấm nhựa tuân theo phân phối chuẩn, dựa
vào dữ liệu hãy kết luận về khẳng định của công ty với mức ý nghĩa α = 5%.

In [8]:
Nhua=pd.read_csv('Nhua.txt')
print(Nhua.head())

   DoBen
0   30.1
1   32.7
2   22.5
3   27.5
4   27.7


Gọi m là khả năng chịu lực trung bình của các tấm nhựa sản xuất bởi công ty này.

H0: m >= 30

H1: m < 30

p-giá trị = 0.042 < 0.05 => Bác bỏ H0 => Ở mức ý nghĩa 5% ta kết luận khẳng định của công ty là không đúng.

In [9]:
# Lệnh tìm p-giá trị
ttest_1samp(Nhua.DoBen,popmean=30,alternative='less')

TtestResult(statistic=-1.8466293347032539, pvalue=0.042312203258501976, df=15)

#### 2. Kiểm định 1 tỷ lệ
Lệnh thực hiện:

In [10]:
# chi tiết các tham số
from statsmodels.stats.proportion import proportions_ztest
?proportions_ztest 

In [23]:
# các tham số
proportions_ztest 

<function statsmodels.stats.proportion.proportions_ztest(count, nobs, value=None, alternative='two-sided', prop_var=False)>

#### Hướng dẫn cách dùng các tham số
##### proportions_ztest(count,nobs,value=None,alternative='two-sided')

Trong đó, count: số quan sát thành công

nobs: Cỡ mẫu

value: giá trị cần so sánh với tỷ lệ tổng thể p trong H0

alternative: {'two-sided', 'smaller', 'larger'}: mô tả giả thuyết đối

Problem 20. Theo số liệu trước đây, dây chuyền sản xuất tại một nhà máy có tỷ
lệ phế phẩm là 7%. Người ta muốn kiểm tra xem tỷ lệ này có thay đổi không nên
đã chọn ngẫu nhiên 250 sản phẩm đem kiểm tra thấy có 20 sản phẩm bị loại. Hãy
tiến hành bài toán kiểm định tại mức ý nghĩa α = 1%.

Giải: Gọi p là tỷ lệ phế phấm của nhà máy ở hiện tại

H0: p = 0.07

H1: p # 0.07

p-giá trị = 0.56 > 0.01 => Chấp nhận H0
=> Ở mức ý nghĩa 1% ta kết luận tỷ lệ phế phẩm của nhà máy không thay đổi.

In [16]:
# Lệnh bài Bài 20
from statsmodels.stats.proportion import proportions_ztest
proportions_ztest(count=20,nobs=250,value=0.07,alternative='two-sided')

(0.5828155051501958, 0.5600175059745569)

Kết quả trả ra:
0.5828155051501958 (số viết trước) là giá trị kiểm định z

0.5600175059745569 (số viết sau) là p-giá trị

Problem 21. Dùng file dữ liệu ThoiGianCho.csv kiểm định giả thuyết tỷ lệ lượt khách
hàng phải chờ trên 7 phút không vượt quá 25%, tại mức ý nghĩa α = 5%.

Giải: Gọi p là tỷ lệ lượt khách hàng phải chờ trên 7 phút.

H0: p <= 0.25

H1: p > 0.25

p-giá trị = 0.4098 > 0.05 => Chấp nhận H0 => Ở mức ý nghĩa 5% ta kết luận tỷ lệ lượt khách hàng phải chờ trên 7 phút không vượt quá 25%.

In [6]:
#Lệnh thực hiện Bài 21
TGC=pd.read_csv('ThoiGianCho.csv')
print(TGC.head())

   ThoiGian
0       1.6
1       6.6
2       5.6
3       5.1
4       3.9


In [8]:
np.sum(TGC.ThoiGian>7) # tìm count

26

In [7]:
len(TGC.ThoiGian) # tìm nobs

100

In [11]:
proportions_ztest(count=26,nobs=100,value=0.25,alternative='larger')

(0.22798037629377677, 0.4098307481416089)

#### 3. Kiểm định 2 trung bình
##### 3.1 Hai mẫu độc lập

In [33]:
# Chi tiết các tham số
from scipy.stats import ttest_ind
?ttest_ind

In [27]:
ttest_ind # các tham số

<function scipy.stats._stats_py.ttest_ind(a, b, axis=0, equal_var=True, nan_policy='propagate', permutations=None, random_state=None, alternative='two-sided', trim=0)>

In [ ]:
# a, b: array-like, hai mẫu
# equal_var: True (hai tổng thể có phương sai bằng nhau)
#            False (hai tổng thể có phương sai khác nhau)
# nan_policy: cách xử lý dữ liệu NaN (giống trong ttest_1samp)
# alternative: mô tả giả thuyết đối (giống trong ttest_1samp)

Bài 23: Trong một nghiên cứu về tác dụng của Vitamin C trong điều trị bệnh
cảm lạnh, 22 người bị cảm lạnh tình nguyện được chia làm hai nhóm: nhóm 1 gồm
10 người được sử dụng một ngày 4 viên thuốc chứa 1 gram vitamin C. Nhóm 2 gồm
12 người còn lại cũng dùng thuốc bề ngoài tương tự nhưng không có vitamin C và
vô hại. Thời gian bị cảm (đơn vị: ngày) của những người này được ghi lại trong file VitaminC.xls. Tại mức ý nghĩa α = 5%, hãy xét xem việc sử dụng 4 gram vitamin C mỗi ngày có làm giảm thời gian bị cảm lạnh hay không? Giả thiết rằng hai tổng thể tuân theo phân phối chuẩn.

In [ ]:
# Gọi m1, m2 lần lượt là thời gian cảm lạnh trung bình của các bệnh nhân có
# dùng và không dùng vitamin C
# H0: m1 - m2 = 0    H1: m1 - m2 < 0
# p-giá trị = 0.03428 < 0.05 => Bác bỏ H0 => Ở mức ý nghĩa 5% ta kết luận uống
# vitamin C mỗi ngày có làm giảm thời gian bị cảm lạnh

In [31]:
vtm=pd.read_excel('VitaminC.xlsx')
print(vtm)

    DungC  KhongC
0     5.5     6.5
1     6.0     6.0
2     7.0     8.5
3     6.0     7.0
4     7.5     6.5
5     6.0     8.0
6     7.5     7.5
7     5.5     6.5
8     7.0     7.5
9     6.5     6.0
10    NaN     8.5
11    NaN     7.0


In [38]:
# Có hai quan sát NaN => nan_policy = 'omit'
# Không có giả thuyết hai tổng thể có phương sai bằng nhau => equal_var = False
ttest_ind(vtm.DungC,vtm.KhongC,nan_policy='omit',alternative='less', 
          equal_var = False)

Ttest_indResult(statistic=-1.9252504477681334, pvalue=0.03428345531776966)

Problem 24. Một kỹ sư xây dựng muốn so sánh sức chịu nén của hai loại bê tông.
Anh ta chọn ra một mẫu gồm 10 tấm bê tông loại thứ nhất, dữ liệu về sức chịu nén (đơn vị: psi) của những tấm bê tông này như sau:

3250 3268 4302 3184 3266 3297 3332 3502 3064 3116

Một mẫu ngẫu nhiên khác gồm 10 tấm bê tông loại thứ hai được chọn và có sức
chịu nén như sau

3094 3106 3004 3066 2984 3124 3316 3212 3380 3018

Giả sử hai tổng thể tuân theo phân phối chuẩn có phương sai bằng nhau. Hãy kiểm
định ở mức ý nghĩa α = 5% xem khả năng chịu lực của hai loại bê tông có khác
nhau hay không?

In [ ]:
# Gọi m1 ,m2 lần lượt là sức chịu nén trung bình của bê tông loại 1 và loại 2
# H0: m1 - m2 = 0     H1: m1 - m2 # 0
# p-giá trị = 0.0722 > 0.05 => Chấp nhận H0
# Ở mức ý nghĩa 5% ta kết luận khả năng chịu lực của hai loại bê tông như nhau.

In [40]:
bt1 = np.array([3250, 3268, 4302, 3184, 3266, 3297, 3332, 3502, 3064, 3116])
bt2 = np.array([3094, 3106, 3004, 3066, 2984, 3124, 3316, 3212, 3380, 3018])
# Có giả thuyết hai tổng thế tuân theo phân phối chuẩn => equal_var = True
ttest_ind(bt1, bt2, equal_var = True, alternative = 'two-sided')

Ttest_indResult(statistic=1.9098214757174472, pvalue=0.07222420370089525)

#### 3.2 Mẫu theo cặp

In [13]:
from scipy.stats import ttest_rel
ttest_rel  
#các tham số như đã giới thiệu, chỉ dùng đến a, b, nan_policy, alternative

<function scipy.stats._stats_py.ttest_rel(a, b, axis=0, nan_policy='propagate', alternative='two-sided', *, keepdims=False)>

Problem 28. Để so sánh số hành khách đi xe bus trong giờ cao điểm buổi sáng
và buổi chiều ở một thành phố, trong một ngày người ta chọn ngẫu nhiên 9 chiếc
xe với lộ trình khác nhau và thu thập số liệu về số hành khách đi xe lúc 7h45’ và 16h45’ (Dữ liệu có trong file SoKhach.csv). Có sự khác biệt về số hành khách trên các chuyến xe giữa buổi sáng và chiều không? Kiểm định ở mức ý nghĩa α = 5% cho biết số hành khách đi xe tuân theo
phân phối chuẩn.

In [ ]:
# Gọi m1, m2 lần lượt là số khách trung bình trong giờ cao điểm buổi sáng và
# buổi chiều
# H0: m1 - m2 = 0    H1: m1 - m2 # 0
# p-giá trị = 0.772 > 0.05 => Chấp nhận H0
# Ở mức ý nghĩa 5% ta kết luận không có sự khác biệt về số khách buổi sáng và 
# buổi chiều

In [43]:
sk=pd.read_csv('SoKhach.csv')
print(sk)

   Sang  Chieu
0    43     41
1    51     49
2    37     44
3    24     32
4    47     46
5    44     42
6    50     47
7    55     51
8    46     49


In [44]:
ttest_rel(sk.Sang,sk.Chieu,alternative='two-sided')

TtestResult(statistic=-0.29981267559834457, pvalue=0.7719622494400781, df=8)

#### 4. Kiểm định 2 tỷ lệ

In [48]:
from statsmodels.stats.proportion import proportions_ztest
proportions_ztest # các tham số
# count: array-like, chứa số quan sát thành công của 2 mẫu
# nobs: array-like, chứa cỡ của 2 mẫu

<function statsmodels.stats.proportion.proportions_ztest(count, nobs, value=None, alternative='two-sided', prop_var=False)>

Problem 31. Từ dây chuyền thứ nhất kiểm tra 110 chi tiết người ta loại đi 30 chi
tiết, từ dây chuyền thứ hai kiểm tra 150 chi tiết người ta loại đi 40 chi tiết. Ở mức
ý nghĩa α = 5% có thể cho rằng chất lượng sản phẩm của hai dây chuyền là như
nhau không?

In [ ]:
# Gọi p1, p2 lần lượt là tỷ lệ chi tiết bị loại của 2 dây chuyền
# H0: p1 = p2
# H1: p1 # p2
# p-giá trị = 0.913 > 0.05 => Chấp nhận H0
# Ở mức ý nghĩa 5% ta kết luận chất lượng sản phẩm của 2 dây chuyền là như nhau.

In [51]:
proportions_ztest(count=[30,40],nobs=[110,150],alternative='two-sided')

(0.10884760595201654, 0.9133233640161736)

Ví dụ 2: Dùng mẫu SoLieu.csv hãy kiểm đinh giả thuyết tỷ lệ nam ở khu vực Thành phố cao hơn tỷ lệ này ở Nông thôn. Dùng mức ý nghĩa 5%.

In [14]:
sl=pd.read_csv('SoLieu.csv')
print(sl.head())

   STT GioiTinh  Tuoi    KhuVuc  ThuNhap
0    1      Nam    30  NongThon       30
1    2       Nu    27  ThanhPho       40
2    3       Nu    32   MienNui       25
3    4      Nam    31  NongThon       35
4    5       Nu    41    HaiDao       42


In [59]:
# Gọi p1, p2 lần lượt là tỷ lệ nam ở 2 khu vực Thành phố và Nông thôn
# H0: p1 <= p2; H1: p1 > p2
# p-giá trị = 0.187 > 0.05 => Chấp nhận H0
# Ở mức ý nghĩa 5% ta kết luận tỷ lệ nam ở Thành phố 
# không cao hơn tỷ lệ này ở nông thôn.

In [55]:
pd.crosstab(sl.GioiTinh,sl.KhuVuc) # tìm được count=[26, 10]

KhuVuc    HaiDao  MienNui  NongThon  ThanhPho
GioiTinh                                     
Nam            6       13        10        26
Nu             6       13        10        16

In [57]:
sl.KhuVuc.value_counts() # tìm được nobs=[42, 20]

ThanhPho    42
MienNui     26
NongThon    20
HaiDao      12
Name: KhuVuc, dtype: int64

In [58]:
proportions_ztest(count=[26,10],nobs=[42,20],alternative='larger')

(0.8880108606111126, 0.18726745465224515)